feat(JS rules): add CWE-94 AWS Lambda code injection

elsapet opened the pull request:
## Description
Add rule to check for code injection using node.js vm module (virtual machine)

Relates to #606 

<!-- Add this section if required
## Related
-->
<!-- Closes some existing issue
- Close #AAA
<!-- References some existing PR
- #CCC
-->

## Checklist

- [ ] I've added test coverage that shows my fix or feature works as expected.
- [ ] I've updated or added documentation if required.
- [ ] I've included usage information in the description if CLI behavior was updated or added.
- [ ] PR title follows [Conventional Commits](https://www.conventionalcommits.org/) format

cfabianski commented on packages/js/README.md:
````diff
@@ -41,6 +41,8 @@ class MyApp {
 ```
 
+**:warning: `@bearer/js` is written in ES6**. If you want to support older browser (IE11) you'll have to add extra configuration to your bundling system. You'll need to explicitly tell your bundler (ex: webpack) to process and transform bearer's files.
+
````
Can we provide an example?

tanguyantoine replied:
we can. we would have to do it for webpack and other bundlers. maybe a link to the wiki where we can update it without having to update the codebase

tanguyantoine replied:
here it is: https://github.com/Bearer/bearer/wiki

tanguyantoine commented:
superseeded by #727

elsapet committed:
feat: add CWE-94 AWS Lambda code injection

didroe commented on pkg/commands/process/settings/rules/javascript/aws_lambda/code_injection.yml:
```diff
@@ -0,0 +1,46 @@
+    filters:
+      - variable: METHOD
+        values:
+          - Script
```
Should we get rid of the variable as there's only one?

didroe reviewed (commented):
We have `javascript_express_eval_user_input` which is similar but there's no overlap. It would be good to use the same patterns (except the user input). Or maybe just have a single `lang` rule with it all in.

Also that express rule uses CWE-95 which is more specific. You could use that one also if it fits.

elsapet committed:
fix: update CWE id and add more eval patterns

elsapet commented:
Thanks @didroe - I've added the patterns and updated the CWE number. 
I'm on the fence about combining them into one `lang` rule as the inputs are specific to Express and AWS lambda, and so maybe they do belong in separate folders. Nonetheless, happy to combine them if you feel strongly 👍

didroe commented on pkg/commands/process/settings/rules/javascript/aws_lambda/code_injection.yml:
```diff
@@ -0,0 +1,63 @@
+patterns:
+  - pattern: |
+      vm.$<METHOD>($<...>$<RAW_INPUT>$<...>)
+    filters:
+      - variable: METHOD
+        values:
+          - createContext
+          - runInContext
+          - runInNewContext
+          - runInThisContext
+          - compileFunction
+      - variable: RAW_INPUT
+        detection: javascript_aws_lambda_code_injection_user_input
+  - pattern: |
+      new vm.Script($<...>$<RAW_INPUT>$<...>)
+    filters:
+      - variable: RAW_INPUT
+        detection: javascript_aws_lambda_code_injection_user_input
```
We could add these to the express rule

didroe reviewed (approved):
I'm happy to leave as two separate rules. 

After writing lots of rules, I think the way we're organising the rules isn't really up to the job. We do a mix of either duplicating parts of rules (maintenance burden, risk of differences like in this case, etc.), or we have to violate the structure and have single rules covering multiple things (rules become large, not easy to understand why something is in a rule, etc). But not something to be dealt with right now

elsapet committed:
chore: update snapshots
